### DISCUSSION TITLE
Enable code ownership without automatic PR notifications for entire teams

### DISCUSSION THREAD

**Discussion by raphaelleu** [community/community#35673] [Pull Requests]
I'm not sure there's another place for feature requests, so here it goes. The problem is that we would like to use code ownership to enforce reviews by certain teams on certain PRs before merging but not have the entire team get notifications for all the updates on a PR. This happens because the owning team/s get auto-assigned as reviewers once the PR gets created and are therefore subscribed to the PR. This creates lots of spam. 

 It would be great if we had the option to use code owners for indicating that certain PRs must be reviewed by specific team members, depending on their contents, but not automatically assign the teams as reviewers or at least not automatically subscribe them to updates on that PR. 

I'm aware that we can all manage our notifications, use draft PRs and I'm also aware of the auto assign functionality, but unfortunately these don't quite fit our workflows. What do you think?

Thanks! 

Edit: after posting this, I noticed there's a similar discussion going on here: https://github.com/orgs/community/discussions/15747. The main post is not exactly the same, but the top comment essentially states the same issue as this.

**Comment by simondelphia** [comment]
You can currently disable the option to require review from code owners. Is that not what you're asking for?

You can also specify code owners for specific folders / subfolders of the repo as opposed to just global.

**Reply by raphaelleu** [reply]
No @simondelphia, see this part: "...we would like to use code ownership to enforce reviews by certain teams on certain PRs before merging...."

**Reply by simondelphia** [reply]
What would that look like?

**Reply by mircea-baja-sophos** [reply]
> What would that look like?

On a branch protection rule you would have a checkbox "Don't notify email owners". This would support a number of scenarios e.g.:
- On a codebase there is large number of code owners. For the target branch the rule has "Require review from code owners" ticked and "Don't notify email owners" ticked. The developer that raises a PR to merge into that branch has to request a review from a few of the code owners involved, the rest are not spammed with email notifications.
- A small team uses PRs on some test automation branches. For the test automation branches the rule has "Require a number of reviewers" set to 2, "Require review from code owners" unticked and "Don't notify email owners" ticked. A developer that raises a PR for the test automation branch has to request a review from 2 team members, but the code owners don't get notified about changes that are relevant to just the test automation team branches.

**Comment by billyjbryant** [comment]
I second this request. There are times when we would like to force a required review by a member of the Security team on certain PRs but the Security Team shouldn't be Code Owners or get notifications for all PRs to the Repo, just when they are added as a required reviewer.

**Comment by dominicfraser** [comment]
It would be great to see this be configurable.

I would be great to have it enforced that a CODEOWNER review is required, and can be manually requested, but that the owner is not auto emailed and auto subscribed to the PR. When multiple PRs are being raised a day to a codeowner the amount of emails due to being assigned, new commits being pushed to each PR, discussions on multiple PRs, and the default reaction is to turn off all notifications at a repo level. This means people move to tagging people in Slack to review, or to respond to comments, which is not a desired workflow.

I came here from https://stackoverflow.com/questions/72534373/how-to-enforce-review-from-codeowners-without-automatically-requesting-a-review

which describes:

> the desired flow for a contributer is something like this:
> open a pull request -> mess around, make changes -> manually request a review when ready (can't merge without codeowner approval)

**Comment by matthew-kaye** [comment]
Would also love to see this. I'm getting bombarded with notifications at the moment, but the only solution is to disable email notifications and risk missing notifications that are actually relevant when I'm assigned!

**Comment by luxaritas** [comment]
Just ran across this issue - there is an individual in our "maintainers" group who should have the ability to sign off PRs, but shouldn't necessarily always be notified as new PRs go up.

**Comment by JustinGol** [comment]
Another vote to have this functionality. I want to protect a certain branch and require codeowners to approve but I dont want them assigned to the PR's on other branches.

**Comment by cmkoller** [comment]
Also want to put in a request for this functionality! To be more specific about the need we've encountered at my organization, we want to use codeowners for branch protection, but don't want the automatic assignment of codeowners to review, and also (especially!) don't want the PR notifications for the whole codeowner team. 

Our workflow would look like this: a member of a non-codeowner team makes a PR, discusses during stand-up with codeowners, a codeowner self-selects to review the PR.

**Comment by adamsuskin** [comment]
<img width="1207" alt="Screenshot 2023-03-23 at 10 38 21 AM" src="https://user-images.githubusercontent.com/5596609/227273420-5663d1bb-db0d-4bcc-a1b2-58d2e37a021e.png">

^ The root cause of my GitHub notification volume issue isn't watching a bunch of no star, inactive repos. It's the problem outlined in this thread.

**Comment by jcook-uptycs** [comment]
You can use this to auto-assign for the team. https://docs.github.com/en/organizations/organizing-members-into-teams/managing-code-review-settings-for-your-team so not everyone get the notice. 

In the same place there is also `Only notify requested team members`

**Comment by parhamfh** [comment]
While auto-assign works I would appreciate more granular control when creating a PR as well. 
If the team is auto-requested through CODEOWNER:
- Setting for sending only initial or no notification for PR being created. This would decrease the constant barrage of emails.
- If the PR author adds members from the CODEOWNERs to the PR at creation time, do not _also_ add the team since CODEOWNERs are represented. This is in line with the team being removed as requested once a member of CODEOWNERs leaves a review. This enables dialogue between PR authors and select CODEOWNERs for reviews, without bothering remaining CODEOWNERs.

**Comment by justin-newman** [comment]
I also agree with this. Right now if CODEOWNERS exists and a developer creates a PR into another developers working branch a code review is requested by the entire dev team (assuming the team owns the code that is being changed), but you would really only want that one dev to be requested. Then once those changes need to be merged into main or some other gitflow branch it would make sense to make the request to the whole team.

**Reply by torressam333** [reply]
@justin-newman did you find a workaround?

**Comment by vbabenkoru** [comment]
I'm getting spammed with dozens of notifications every day for various PRs, even for repositories explicitly added to the ignore list – I still keep getting notifications from those with no way to disable. The only way currently is to get removed from the group listed in CODEOWNERS.

**Comment by thiagobdp** [comment]
We have the same issue. We are automatically subscribed to every PR and receive infinite spam from GitHub. 
Result: I had to filter GitHub emails to trash. It's totally useless...

**Comment by JacobCoffee** [comment]
We've had members leave because their team is assigned to PRs but the individuals didn't want 875 messages in their email or GH notification page. +1

**Comment by benb111** [comment]
Also have the same issue here. We are using CODEOWNERS to implement branch protection on a subset of the files in the repo, but I don't want everyone getting an email for every PR in the repository.

**Comment by torressam333** [comment]
Any solution to this?

**Comment by dgoldstein0** [comment]
Also had this problem. 

My workaround was to disable email notifications entirely and configure slack notifications instead. One of the biggest downsides with the slack notifications is that they only notify me when my diffs land, not diffs I review, so still leaves a gap between what I get notified of and what I want to know about. 

My company is actually building our own notification service because of these issues.  Which is ridiculous because I'm sure we pay a lot for github, and clearly the non configurability of the email notifications is a problem for lots of medium sized and larger companies.

**Reply by torressam333** [reply]
What a bummer. I had to delete CODEOWNERS and then add a custom workflow that requires a codeowner approval without explicit email notifications.

**Reply by dgoldstein0** [reply]
We've also built our own ownership reviews check that we require for merging instead of github's.  I think that was driven by other concerns though.  We're using that and code owners because code owners is the only thing that gives us ui that shows who owns what or let's you filter to changes to your team's files

**Comment by IanGrainger** [comment]
It seems like if you have a team of more than (say) two people, almost no-one would want ALL of the team automatically subscribed to ALL of the PRs so they get notified of ALL of the comments.

> What a bummer. I had to delete CODEOWNERS and then add a custom workflow that requires a codeowner approval without explicit email notifications.

This sounds like a workaround, how did you do this please?

**Comment by elias-pap** [comment]
> You can use this to auto-assign for the team. https://docs.github.com/en/organizations/organizing-members-into-teams/managing-code-review-settings-for-your-team so not everyone get the notice.
> 
> In the same place there is also `Only notify requested team members`

We use this plus a dummy user that gets auto-assigned all the PRs and notifications by default, so no one else gets them. It looks like this:

1. Add a dummy user to your team.
2. Go to https://github.com/orgs/your-org/teams/your-team/edit/review_assignment
3. In "Never assign certain team members", select **all team members** other than the dummy user.
4. Configure like this:
<img width="1047" alt="codereview-workaround" src="https://github.com/user-attachments/assets/d6f1888f-d2c7-449f-8409-fdd08d59c353">

**Comment by dstorch** [comment]
This is an issue for my company as well! We have a large engineering organization, so each of our Github teams may consist of 10 or even 30 engineers. When those larger teams are assigned as CODEOWNERS, the volume of emails gets too large. We would love to see more configuration options around email notifications for PRs.

The workaround we have implemented is similar to what @elias-pap described above:
* We created a dummy account.
* We have a script in our build process that generates the CODEOWNERS file, and this makes the dummy account an explicit owner of everything (in addition to the actual owning team).
* Teams who don't want email notifications on PRs can then add the dummy account to their Github team and can configure the "Only notify requested team members" option. Since the dummy account is always requested for review (it owns everything), and since the dummy account is a member of the team, the team will not be notified.

**Reply by Muneeb147** [reply]
But that will always require dummy's account approval. The original team members approval would be of no use unless or until the dummy account approves.

I am assuming you are talking to have following entry in CODEOWNERS file:

```
* @dummy-acount
* @actual-team
```

**Reply by dstorch** [reply]
@Muneeb147 our CODEOWNERS file looks like this:

```
file.cpp @actual-team @dummy-account
```

My understanding is that this means that either a member of the actual team or the dummy account must approve. Approvals from both are not required.

**Reply by Muneeb147** [reply]
Yes this structure makes sense.

**Reply by cburroughs** [reply]
Are you seeing this 'trick' work with the Slack integration as well?  (What our team is reporting  so far is that email seems to work but notification for slack seems to go through regardless.)

**Comment by gauthierpetetin** [comment]
This is my biggest pain point on Github.

- I need to be notified when I'm personally @mentionned.
- I don't want to be notified for all activity on PRs where my team is listed as reviewer, as this happens hundreds of time per day.

Due to the latter, I currently receive hundreds of notifications per day, which constantly disrupts me in my work. I can't blindly delete all these notifications because I first need to check for each notification if it's destined to me personally or if it's destined to my team.

**Comment by adamsuskin** [comment]
Hi everyone, I've reached out to some folks here already but wanted to share here too:

My team [Upshift](https://upshift.dev/) is now working on a replacement for codeowners (migration tool included) that can prevent PR authors from notifying the owners. This can be used as a replacement workflow so you can keep your inboxes much cleaner!

It's called [code-captains](https://github.com/upshift-dev/code-captains), and it's open-source and MIT-licensed. Please feel free to reach out on Github or to me personally if you want to talk about code ownership / request more features.

**Comment by MickeyPagaya** [comment]
This is indeed really needed, we've reduced our use of code ownership just because of this reason.

**Comment by cunisia** [comment]
It's crazy that this issue is still not fixed. Why can't we configure branch protection so that review is required from people belonging to some specific team just like in gitlab ? All we want is a rule to be able to merge, not auto assignation of PR that will lead to notifications.

**Comment by mateusz-wlosek-ppro** [comment]
My company is affected as well. Will Github ever do anything about such issues or will they keep ignoring all the issues with user experience in Github?

**Comment by RajuKattamanchi** [comment]
I second this problem and wish GitHub can do something about it.

Also, The CODEOWNERS file doesn’t have an option to differentiate between draft and ready PRs. I do not want my code owners notified if its just draft PR.

**Comment by gdrt94** [comment]
I'm currently dealing with this using Gmail filters.

I have noticed that Github adds the following footer to such emails: "You are receiving this because your review was requested". But when someone explicitly mentions you, not just the team, the footer is "You are receiving this because you were mentioned".

So I have created the following filter in gmail to filter out github noise:
`Matches: from:(notifications@github.com) "You are receiving this because your review was requested" Do this: Skip Inbox, Apply label "Github Noise"`

Everything else from Github ends up in my inbox.
Hope this helps.

**Reply by CreepyGnome** [reply]
That footer message is a 1 to 1 mapping to just using the email `review_requested@noreply.github.com` being CC'd to, which is more efficient way to filter over a string match in the body of a message.  However that doesn't solve the issue it is actually hiding the problem.  That is how we should be notified that where were requested, but with codeowers it breaks that if you have 100 code owners all of them will get an email requesting their review.

I think @elias-pap  suggestion above where actually is a good hybrid solution to use with codeowners. I was able to get it working and it isn't a perfect solution it does work just not convenient or obvious and requires you to be a maintainer of the group that is the codeowner, etc etc. But still it is a viable option.
https://github.com/orgs/community/discussions/35673#discussioncomment-10340196

**Comment by CreepyGnome** [comment]
Hmm so it seems like GitHub does not provide a way to limit who is ALLOWED to approve PRs to a specific group of people, without also adding them automatically to the PR for review and then sending spamming all them an email.

It would be nice is there was a way to have two groups of people, the superset that has ALL the people allowed to approve PRs and then a subset of those that are automatically added to the PR for as a group who would then be notified.  This way we do not spam 100's of people, when we only need to spam maybe 10-20, as we have a number of people who should be allowed to approve them to pass checks but do not need nor want to get the email notifications.

A quick and simple use case is let say you have staff engineers, principle engineers, architects and they need and should ahve the ability to approve PRs and pass the check that they are approved. While not spamming them with every PR a repo makes.  They would like to know when someone has manually added them for their opinion or desire for feedback, so they can know when they see an email from `review_requested@noreply.github.com` it really means something to them and they can prioritize it to be reviewed.

The current implementation in GitHub if we use CODEOWNERS file is that we will be spamming everyone in this file by that email address which will make that email address worthless to those as a tool to filter by.

Not sure if there is another way with existing GitHub settings or CODEOWNERS to achieve this tiered approach for approvers and those notified automatically to review a PR.

**Reply by cunisia** [reply]
For your specific use case @elias-pap work around might work 
https://github.com/orgs/community/discussions/35673#discussioncomment-10340196

**Reply by CreepyGnome** [reply]
Thanks, I can look into that, but it will require lots of back-and-forth with the org admins, and since this will be a trial-and-error thing, I'm not sure if they will want to do it. I will try to give it at least a shot. lol

**Comment by belm0** [comment]
> I have noticed that Github adds the following footer to such emails: "You are receiving this because your review was requested".

This doesn't distinguish between you being specifically assigned for review vs. team being assigned for review.

**Reply by gdrt94** [reply]
Since we're not using personal review assignment functionality in our team, this is fine.
All I care is that when someone comments on my PRs, replies to my comments, or mentions me in someone else's PRs, I get a notification and that in all other cases I don't get a notification.

**Comment by dvochoa** [comment]
This issue has caused me so much frustration, please fix this 😭

**Reply by CreepyGnome** [reply]
Have you tried @elias-pap suggestion above
https://github.com/orgs/community/discussions/35673#discussioncomment-10340196

**Comment by codymoorhouse** [comment]
This would be a fantastic feature, would love to prevent automatic PR notifications for the entire team but still assign code ownership

**Comment by marcoieni** [comment]
GitHub, consider adding this feature to the codeowners file:

- if the file contains `@username`, they get notified (current behavior)
- if the file contains `@_username` they don't get notified but they can still approve changes to those files.

**Comment by corsonknowles** [comment]
Can we ship this already, please?

**Comment by waltervillarreal** [comment]
I second this, its really annoying

**Comment by cgruber** [comment]
This is starting to look like Github is abandonware. (I know that's hyperbolic, but this dialogue is so similar to abandoned projects)

**Comment by epiqu1n** [comment]
Please add this! The amount of email noise has _immediately_ become problematic since we added code ownership rules for our team's files, but the benefit of code ownership rules is too much to give up.

At our org, if a team needs external code review then they just drop a message in the other team's channel. We don't need the entire team getting emails for every PR that goes up where we're code owners, let alone emails for every comment and update that happens on all of those PRs.

It would be a big quality of life improvement to be able to turn that off, and it's a bit frustrating that we are _so_ close to it given there is already the "Only notify requested team members" setting (which only works if you set up auto-assignment).

**Comment by tarfeef101** [comment]
This is so crucial I can't believe it isn't already a thing, and doesn't have thousands of upvotes.

My team has ownership of tons of things. However, we assign 1 person on any given day to be the supporting engineer who fields questions, reviews PRs, etc. We want to express ownership through CODEOWNERS to ensure someone from our team reviews changes to stuff we own, but we neither need nor want the entire team to get emails about every single one of these changes. We have external mechanisms to figure out who specifically will review something (manually adding someone who worked with you on something, or was the original author, "tagging that person handling support tasks for the day", usually all this is through a messaging tool like slack...). github should be more flexible to allow for teams' workflows.

**Comment by Mariuble** [comment]
Still waiting. Our workaround is to not use codeowners...

**Comment by Cordazar** [comment]
GitLab solves it like this: https://docs.gitlab.com/user/project/codeowners/reference/#optional-sections

the [Go] section is optional:
```
[Documentation]
*.md @root

[Ruby]
*.rb @root

^[Go]
*.go @root
```

Should be easy enough to implement with the resources of M$ at your disposal, right?

**Comment by mluogh-xdof** [comment]
This cannot be that hard to implement. 
Any brave souls?
<img width="640" height="356" alt="image" src="https://github.com/user-attachments/assets/fc7ec497-0e33-4a94-a655-fc58c3979c7f" />

**Reply by tarfeef101** [reply]
Wasn't for this but I have considered that in the past 🤣

**Comment by beatngu13** [comment]
This might be a workaround to enforce code ownership without having to deal with all the notifications:

> # New ruleset bypass type
> 
> GitHub repository rulesets now let you exempt specific actors (users, teams, or GitHub Apps) from all enforced rules with a new `exempt` bypass type. Unlike a standard bypass (i.e., a “break glass” action that requires an explicit actor bypass and generates prominent audit signals), an exemption silently skips enforcement. Use it for trusted, high-volume automation or migration scenarios where friction would otherwise slow you down.
> 
> ## Key benefits
> 
> * Reduce friction for trusted high-volume automation (migration or integration bots)
> * Speed large repository migrations without disabling rules
> * Match classic branch protection behavior for “Restrict who can push”

https://github.blog/changelog/2025-09-10-github-ruleset-exemptions-and-repository-insights-updates/#new-ruleset-bypass-type

Create a ruleset, select your team's repositories (e.g. via custom properties) and restrict updates. If you now add your team to the bypass list as an exempt, the team basically controls the updates being made.

**Reply by theophanemayaud** [reply]
Was looking at this but can't find a way to require pull request approvals from people defined in the ruleset bypass. I would need to have a branch protection rule that blocks approvals, and then allow people in the bypass to be able to approve. But as I see it, that's not possible, it's only possible to block all contributions except those in the bypass, not to block only approvals... Or did you find a way to?

**Reply by beatngu13** [reply]
@theophanemayaud yes, everyone can still approve, but it has no effect unless someone from the bypass list approves. Here is the JSON:

```json
{
  "name": "code-ownership",
  "target": "branch",
  "enforcement": "active",
  "conditions": {
    "ref_name": {
      "exclude": [],
      "include": [
        "~DEFAULT_BRANCH"
      ]
    }
  },
  "rules": [
    {
      "type": "update"
    }
  ],
  "bypass_actors": [
    {
      "actor_id": 123,
      "actor_type": "Integration",
      "bypass_mode": "exempt"
    },
    {
      "actor_id": 456,
      "actor_type": "Team",
      "bypass_mode": "exempt"
    }
  ]
}
```

The first exempt is a GitHub app (Renovate) for automerges, the other one is the team owning the repository. Only if these actors approve, the PR can be merged.

**Comment by wibewave2516-alt** [comment]
A city painting captures the charm, movement, and energy of urban life in a timeless work of art. From modern skylines to historic streets, each[ city painting](https://www.home-original.eu/product-category/cityscape-paintings/) tells a unique story that adds depth and sophistication to your home décor. Hand-painted with precision and creativity, these artworks are perfect for living rooms, hallways, and offices, creating an inspiring atmosphere. Our collection of city paintings blends contemporary style with classic urban themes, making them versatile for any interior. Discover the beauty of city painting and bring the essence of your favorite cities into your living space today.

**Comment by staffann** [comment]
The lack of this feature (require codeowners review without automatically adding all code owners to a PR) is a major source of displeasure at the company I work for. Please fit it!

**Comment by jwtracy** [comment]
It's absurd that this many people can ask for this basic functionality for years without GitHub providing any update or workaround

**Reply by theophanemayaud** [reply]
Most people at my org have just made all GitHub emails go to trash: this is a serious problem, preventing actual use of all other good GitHub emails ! 
I resist by adding layers upon layers of email filters, but has to be done per repo, and is tedious to manage while still making some useful notifications go unoticed!

**Comment by w-biggs** [comment]
My workaround for now is to have all emails from Github for certain repositories automatically go to a folder where I can ignore them, with exceptions for if the `X-Github-Reason` header is `author` or if the `X-Github-Sender` header is my username. There are probably a few emails I would want to get that also get caught in this Outlook rule but it's worth it. It's ridiculous that there's no better way to do this

**Reply by dserodio** [reply]
This may be the first time I envy Outlook users, since it's not possible to filter by arbitrary headers in Gmail.

**Comment by bibimorlet** [comment]
+1 for this feature request.

At our organization, we rely on CODEOWNERS for branch protection rules, but we explicitly want to avoid automatic reviewer assignment.
The main pain point is the notification noise: tagging entire CODEOWNER teams on every PR creates a lot of unnecessary alerts, especially for large teams.

Being able to enforce ownership without auto-assigning reviewers or notifying teams would be a big improvement.

**Comment by dtitov** [comment]
We really need it in our team, please take it into work 🙏

**Reply by simukka** [reply]
Yes. My team is tired of my spam. Please help us!

**Comment by LucasAdamsTR** [comment]
Looks like this feature actually does exist and was released a couple of months ago? https://github.blog/changelog/2025-11-03-required-review-by-specific-teams-now-available-in-rulesets/

**Reply by theophanemayaud** [reply]
Nope this also works by adding the team set as required, to be reviewer of the non draft PRs, which results in the same emails. The problem is not the code owners per say, it's not being to distinguish emails from "manually requested team/person as reviewer" from "automatically assigned team/person as reviewer because of required review rules"

**Comment by melanciani** [comment]
Would be great to have an official statement about why this feature has not been considered yet after more than 3 and half years 🙂  It would be very useful to avoid unnecessary spam notifications by default !

**Comment by aharwood2** [comment]
Bump!

**Comment by devgusto** [comment]
+1

**Comment by cemtu** [comment]
+1

**Comment by jjjhill** [comment]
+1

**Comment by bf-luca** [comment]
+1 for this feature.

With the deprecation of Dependabot's "reviewers" feature over a year ago, disabling automatic PR notifications for Codeowners is a must. As of right now, using Dependabot with Codeowners you are getting flooded with updates on PRs.
This issue renders using both Dependabot and Codeowners practically unusable.

**Comment by Metadorius** [comment]
Lack of this feature makes notifications from GH unusable for me and many people I work with.

**Comment by galloviolet** [comment]
+1. use of codeowners is a company policy i don't have the authority to overrule, and this makes my email notifications inbox from github essentially unusable.

**Comment by TrimVis** [comment]
+1

**Comment by stoicaeduardconst** [comment]
+1

**Comment by diogeneshamilton** [comment]
Lack of this feature makes notifications from GH unusable for me and many people I work with.

**Comment by OutSystemsAMM** [comment]
+1, the use of codeowners is a company policy here too. And I get it, it mainly means for us: X team owns repo Y, that's necessary. We have other means and automations of ensuring that PRs don't get lost. We have no use for the automatic assignment of codeowners as reviewers. This could be configurable per repo. We will end up just disabling all notifications.
